File: mobile/lib/models/policy.dart
/// Policy model with full properties and JSON serialization.
class Policy {
  final String id;
  final String policyNumber;
  final String policyholderId;
  final String policyTypeId;
  final String policyTypeName;
  final double coverageLimit;
  final double premium;
  final double deductible;
  final DateTime startDate;
  final DateTime expiryDate;
  final String status;
  final String renewalStatus;
  final String? exclusions;
  final bool isExpired;
  final bool canRenew;
  final List<PolicyCoverage> coverages;
  final DateTime createdAt;
  final DateTime updatedAt;

  Policy({
    required this.id,
    required this.policyNumber,
    required this.policyholderId,
    required this.policyTypeId,
    required this.policyTypeName,
    required this.coverageLimit,
    required this.premium,
    required this.deductible,
    required this.startDate,
    required this.expiryDate,
    required this.status,
    required this.renewalStatus,
    this.exclusions,
    required this.isExpired,
    required this.canRenew,
    this.coverages = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Policy.fromJson(Map<String, dynamic> json) {
    return Policy(
      id: json['id'] as String,
      policyNumber: json['policyNumber'] as String,
      policyholderId: json['policyholderId'] as String,
      policyTypeId: json['policyTypeId'] as String,
      policyTypeName: json['policyTypeName'] as String? ?? '',
      coverageLimit: (json['coverageLimit'] as num).toDouble(),
      premium: (json['premium'] as num).toDouble(),
      deductible: (json['deductible'] as num).toDouble(),
      startDate: DateTime.parse(json['startDate'] as String),
      expiryDate: DateTime.parse(json['expiryDate'] as String),
      status: json['status'] as String,
      renewalStatus: json['renewalStatus'] as String,
      exclusions: json['exclusions'] as String?,
      isExpired: json['isExpired'] as bool? ?? false,
      canRenew: json['canRenew'] as bool? ?? false,
      coverages: (json['coverages'] as List<dynamic>?)
              ?.map((c) => PolicyCoverage.fromJson(c as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'policyNumber': policyNumber,
      'policyholderId': policyholderId,
      'policyTypeId': policyTypeId,
      'policyTypeName': policyTypeName,
      'coverageLimit': coverageLimit,
      'premium': premium,
      'deductible': deductible,
      'startDate': startDate.toIso8601String(),
      'expiryDate': expiryDate.toIso8601String(),
      'status': status,
      'renewalStatus': renewalStatus,
      'exclusions': exclusions,
      'isExpired': isExpired,
      'canRenew': canRenew,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

/// Policy coverage model.
class PolicyCoverage {
  final String id;
  final String policyId;
  final String coverageType;
  final String? description;
  final double coverageLimit;
  final double deductibleAmount;
  final double percentageOfCoverage;
  final bool isActive;

  PolicyCoverage({
    required this.id,
    required this.policyId,
    required this.coverageType,
    this.description,
    required this.coverageLimit,
    required this.deductibleAmount,
    required this.percentageOfCoverage,
    required this.isActive,
  });

  factory PolicyCoverage.fromJson(Map<String, dynamic> json) {
    return PolicyCoverage(
      id: json['id'] as String,
      policyId: json['policyId'] as String,
      coverageType: json['coverageType'] as String,
      description: json['description'] as String?,
      coverageLimit: (json['coverageLimit'] as num).toDouble(),
      deductibleAmount: (json['deductibleAmount'] as num).toDouble(),
      percentageOfCoverage: (json['percentageOfCoverage'] as num).toDouble(),
      isActive: json['isActive'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'policyId': policyId,
      'coverageType': coverageType,
      'description': description,
      'coverageLimit': coverageLimit,
      'deductibleAmount': deductibleAmount,
      'percentageOfCoverage': percentageOfCoverage,
      'isActive': isActive,
    };
  }
}

File: backend/src/InsuranceClaims.Domain/PolicyManagement/Policy.cs
using InsuranceClaims.Domain.Common;
using InsuranceClaims.Domain.PolicyManagement.Enums;

namespace InsuranceClaims.Domain.PolicyManagement;

/// <summary>
/// Represents an insurance policy.
/// </summary>
public class Policy : BaseEntity
{
    public string PolicyNumber { get; set; } = string.Empty;
    public Guid PolicyholderId { get; set; }
    public Guid PolicyTypeId { get; set; }
    public decimal CoverageLimit { get; set; }
    public decimal Premium { get; set; }
    public decimal Deductible { get; set; }
    public DateTime StartDate { get; set; }
    public DateTime ExpiryDate { get; set; }
    public PolicyStatus Status { get; set; } = PolicyStatus.Draft;
    public RenewalStatus RenewalStatus { get; set; } = RenewalStatus.NotDue;
    public string? Exclusions { get; set; }

    // Navigation properties
    public PolicyType? PolicyType { get; set; }
    public ICollection<PolicyCoverage> Coverages { get; set; } = new List<PolicyCoverage>();

    // Business logic

    /// <summary>
    /// Returns true if the policy has passed its expiry date.
    /// </summary>
    public bool IsExpired() => DateTime.UtcNow > ExpiryDate;

    /// <summary>
    /// Returns true if the policy is eligible for renewal.
    /// A policy can be renewed if it is Active or Expired (within grace), and not already renewed/cancelled/lapsed.
    /// </summary>
    public bool CanRenew()
    {
        if (Status == PolicyStatus.Cancelled || Status == PolicyStatus.Lapsed)
            return false;

        if (RenewalStatus == RenewalStatus.Renewed)
            return false;

        return Status == PolicyStatus.Active || Status == PolicyStatus.Expired;
    }

    /// <summary>
    /// Marks the policy as lapsed.
    /// </summary>
    public void Lapse()
    {
        Status = PolicyStatus.Lapsed;
        RenewalStatus = RenewalStatus.Declined;
    }
}
